packages/vite/src/node/plugins/worker.ts

```diff
@@ -276,11 +275,24 @@ export function webWorkerPlugin(config: ResolvedConfig): Plugin {
         }`,
         map: { mappings: '' } // Empty sourcemap to supress Rolup warning
       }
-    },
+    }
+  }
+}
+
+// just run in build mode
+export function webWorkerImportMetaUrlPlugin(config: ResolvedConfig): Plugin {
+  const isWorker = config.isWorker
 
-    renderChunk(code) {
-      if (isWorker && code.includes('import.meta.url')) {
-        return code.replace('import.meta.url', 'self.location.href')
+  return {
+    name: 'vite:workerImportMetaUrl',
+    transform(code) {
+      if (isWorker && config.worker.format !== 'es') {
+        // if build with iife it will polyfill with rollup will be used document in the worker
+        // else if build with es it will replace with `define plugin` will got a unexpected data.
+        // so replace import.meta.url break the default handle.
+        // And it must be done at this (after `(?new Worker)new URL('xxx', import.meta.url)` match),
+        // Otherwise, the following regex will be incorrectly matched.
+        return code.replace(/\bimport.meta.url\b/g, 'self.location.href')
```

Comment:
It is lacking sourcemap support here.

Reply:
https://github.com/vitejs/vite/pull/7706#discussion_r849105361

I don't know why it's better now.

Reply:
I think if there is an `module` variable, esbuild treats that file as `cjs`. Because the `format` is set to `esm`, esbuild tries to transpile `cjs` into `esm`. So it outputs some broken code.
IIUC this is an edge case bug of esbuild.
So I think it is ok to just change this filename to `mod.js`. (any name which is not `module` will be ok)
https://github.com/vitejs/vite/blob/main/packages/playground/worker/modules/module.js

Reply:
> I think if there is an module variable, esbuild treats that file as cjs. Because the format is set to esm, esbuild tries to transpile cjs into esm. So it outputs some broken code.

Thank you very much for your detailed answer. This means that in the case of es, esbuild is also compatible with packaging cjs

Reply:
https://github.com/vitejs/vite/pull/7706#discussion_r849116332

I think issues is in `module` value no filename ?

https://hyrious.me/esbuild-repl/?version=0.14.36&mode=transform&input=+var+module+%3D+%27A+string%27%3B%0A%0Avar+a+%3D+%7B%0A++%27default%27%3A+module%0A%7D%0A&options=--minify+--format%3Desm

Reply:
Yes the issue is in `module` variable. The name of variable comes from the filename. So if you rename `packages/playground/worker/modules/module.js` to `mod.js`, the name of variable will become `mod` instead of `module`.

Reply:
oh sorry I was think resolved before.